Allow filtering workflows to only running, & add dashboard links

emma-sg opened the pull request:
Closes #2527
Closes #2682

## Changes
- Adds backend support for filtering workflows by running state
- Adds "Schedule" filter, replacing "All", "Scheduled", and "No Schedule" tabs
- Persists filter and sort state in URL search params, allowing for linking to a specific filter
- Updates `SearchParamsController` with a more flexible API
  - Renames `set` method to `update`
  - Adds new `set` and `delete` methods that mirror the ones on `URLSearchParams`
- Adds deep links to dashboard

## Testing

Since this PR involves both frontend and backend changes, you'll need to run the backend locally (or deploy this on dev) to test this out.

1. Start a crawl
2. Navigate to the dashboard. Once the crawl has moved past the "starting" phase, verify that the dashboard shows that there's (at least) one running crawl.
3. Click the "→" button next to the "X Crawls Running", and verify that it takes you to the crawling page with the "Show Only Running" filter enabled
4. Click on various other filters (All, Scheduled, etc), and verify that the list and URL update to match
5. Switch between different sorting modes, and verify that the list and URL update to match
6. Using your browser's back and forward buttons, verify that the filters and sorting update when navigating
7. On an org with more than one page of workflows, verify that:
   1. You can navigate through multiple pages of workflows and use filters
   2. Filters don't interfere with each other
   3. Changing filters and sorting resets the page back to the first page
8. Reload the page with some filters and sorting settings set, and verify that the same filters and sorting settings are applied

## Screenshots

| Workflow List Filter | Dashboard Section |
|--------|--------|
| <img width="991" alt="workflow filters" src="https://github.com/user-attachments/assets/c69e39a1-cee7-4042-aed5-bf44b9346c38" /> | <img width="1261" alt="dashboard full" src="https://github.com/user-attachments/assets/f250c938-fe72-463c-8e67-5fc012554aa9" /> | 
| <img width="399" alt="workflow filter detail" src="https://github.com/user-attachments/assets/bc8621ab-1943-4f1e-a6c2-59c278f6f53f" /> | <img width="478" alt="dashboard detail" src="https://github.com/user-attachments/assets/58a0a5c4-ceb0-418e-a27f-1536edb4e178" /> |

| Filters & sorting in URLs |
|--------|
| <img width="382" alt="url 1" src="https://github.com/user-attachments/assets/1fb84f9a-66b2-4ae9-a137-19e6a4b9cfa7" /> |
| <img width="484" alt="url 2" src="https://github.com/user-attachments/assets/d1b99e8f-aefa-4750-a4f6-cef89c08abf4" /> |


https://github.com/user-attachments/assets/362f402c-7e75-464d-b255-0bc7e6145bfa

emma-sg committed:
allow filtering by crawl running state

emma-sg committed:
add set & delete methods to search params controller
- renames `set` method to `update`
- adds `set` and `delete` methods that mirror `URLSearchParams` methods

emma-sg committed:
simplify pagination url search params updates

emma-sg committed:
add "running" filter to crawl workflow list
- add "running" filter
- persist these filters in url search params

emma-sg committed:
update filters from search params on first render

emma-sg committed:
add link to running crawls from dashboard
- add link to running workflows
- remove dt/dd formatting so that underlines can be rendered
- use tabular numbers in overview

emma-sg committed:
re-add dt/dd elements, and use buttons instead
- add more dashboard links as well

SuaYoo commented on frontend/src/pages/org/dashboard.ts:
```diff
@@ -255,6 +255,9 @@ export class Dashboard extends BtrixElement {
                     color: this.colors.crawls,
                   },
+                  button: {
+                    url: `/orgs/${this.orgId}/items/crawls`,
```
Using `this.navigate.orgBasePath` will remove the extra processing of an org ID to slug. Also, `/crawls` is not a valid path, it should be `crawl`--same for `upload`.
```suggestion
                    url: `/${this.navigate.orgBasePath}/items/crawl`,
```

emma-sg replied:
My bad, thank you!

emma-sg committed:
fix incorrect links on dashboard

Co-authored-by: sua yoo <sua@suayoo.com>

SuaYoo commented on frontend/src/pages/org/workflows-list.ts:
```diff
@@ -418,44 +459,77 @@ export class WorkflowsList extends BtrixElement {
-            @click=${() =>
-              (this.filterBy = {
+            @click=${() => {
+              this.searchParams.set("filter", "unscheduled");
```
Since search params are coupled with`this.filterBy`, it should probably be reactive, something like (untested:)

```ts
updated(changedProperties) {
  if (changedProperties.has("filterBy")) {
    this.searchParams.delete("page");

    if (this.filterby.schedule) {
      this.searchParams.set("filter", "unscheduled");
      // etc, more logic
    }
  }
}
```

emma-sg replied:
Yeah, I guess that works, but it feels awkward to me, there's a lot of indirection:
```ts
  protected updated(
    changedProperties: PropertyValues<this> & Map<string, unknown>,
  ) {
    if (changedProperties.has("filterBy")) {
      this.searchParams.delete("page");
      if (
        this.filterBy.schedule === undefined &&
        this.filterBy.isCrawlRunning === undefined
      ) {
        this.searchParams.delete("filter");
      } else if (
        this.filterBy.schedule === true &&
        this.filterBy.isCrawlRunning === undefined
      ) {
        this.searchParams.set("filter", "scheduled");
      } else if (
        this.filterBy.schedule === false &&
        this.filterBy.isCrawlRunning === undefined
      ) {
        this.searchParams.set("filter", "unscheduled");
      } else if (
        this.filterBy.schedule === undefined &&
        this.filterBy.isCrawlRunning === true
      ) {
        this.searchParams.set("filter", "running");
      }
    }
  }
```

SuaYoo replied:
Maybe this indicates that the search params could be more mappable, something like `?scheduled=true/false`, `?running=true/false`? Maybe a simplified transformer since `setFiltersFromSearchParams` feels similarly expanded?

emma-sg replied:
I think I found a pretty good solution in 3024f6d — with a more expanded search param schema I'd also want more UI that maps to that, and I think the current setup functions more like presets than it does a full set of filter controls. I'm not opposed to the idea in the future, but I think it'd take a bit more UI design first.

emma-sg replied:
On second thought, I think I've come around on this, I do see the benefit when it comes to decoupling UI from stored state a bit more. I took a go at this in e83686b, let me know what you think?

emma-sg committed:
simplify filter preset logic

emma-sg committed:
don't push no-change search param updates to history

emma-sg committed:
store filterBy state in searchParams instead of filter preset

emma-sg committed:
wip

emma-sg committed:
Merge branch 'main' into add-only-running-crawl-filter

emma-sg committed:
Merge branch 'main' into add-only-running-crawl-filter

emma-sg committed:
add "mine" to url params

emma-sg committed:
tweak labels and positioning

emma-sg committed:
add query string persistance to sort options

emma-sg committed:
tweaks to layout at small sizes & capitalization

emma-sg committed:
add tooltip, labels, and sort direction to sort button

emma-sg committed:
fix dashboard button link urls

emma-sg committed:
fix a couple bugs & clean up a bit

emma-sg committed:
correctly remove current user filter if not present in params

emma-sg committed:
remove unused const

emma-sg committed:
move filterByCurrentUser initialization to connectedCallback hook

SuaYoo reviewed (changes requested):
The radio button renders over the workflow search filter:

<img width="866" alt="Screenshot 2025-06-26 at 4 17 51 PM" src="https://github.com/user-attachments/assets/0a9d3e56-f4bd-44d5-bfdf-9d502e0ab948" />


Per Discord thread, I made some suggestions for styling and labeling the schedule control in https://github.com/webrecorder/browsertrix/pull/2697, feel free to modify and merge as you see fit.

SuaYoo committed:
Update radio button styles (#2697)

emma-sg committed:
ensure combobox dropdown is rendered above radio buttons

emma-sg committed:
rework layout so that search bar placeholder text is visible for longer,
and less space is left unused in search/filter controls at small-medium
viewport widths

emma-sg commented:
Good catches, thanks! These should both be fixed as of e095877.


https://github.com/user-attachments/assets/11eb66a6-ea93-4400-aa40-2e5f1cb9d20b

tw4l commented:
Looks great! Only thing I've noticed (admittedly an edge case) is that if someone enters/pastes a URL into the browser nav bar with a page number higher than the number of real pages, they end up in a bit of a weird state:

<img width="1180" alt="Screenshot 2025-06-30 at 12 41 15 PM" src="https://github.com/user-attachments/assets/7ea390da-2e47-4cae-b7ae-4b6d55a47032" />

Might be better to just send them to the last page that does have content and update the URL accordingly?

Otherwise everything appears to be working perfectly, and these are great improvements! Well done :)

emma-sg committed:
improve clamping logic

emma-sg commented:
> Only thing I've noticed (admittedly an edge case) is that if someone enters/pastes a URL into the browser nav bar with a page number higher than the number of real pages, they end up in a bit of a weird state

Good catch, added an extra clamping check in 31c941e to address this.

SuaYoo reviewed (approved):
Nice! Once thing I noticed is that the dashboard "View All" button for crawls feels a bit inconsistent when running crawls is 0 and waiting is >0 (as opposed to Collections, where "View all" is always next to the total)

<img width="457" alt="Screenshot 2025-06-30 at 10 53 20 AM" src="https://github.com/user-attachments/assets/8eefe903-ebd5-4902-bf72-950030aa7b0e" />

Not a blocker, just wanted to note for later--I think we can address this later since this is due to how existing dashboard stats are grouped and organized.
